doc/guide/guide-bloch.rst

```diff
@@ -407,7 +407,7 @@ An example of generating images for generating an animation outside of Python is
 
 Generating an animation using ffmpeg (for example) is fairly simple::
 
-   ffmpeg -r 20 -b 1800 -i bloch_%01d.png bloch.mp4
+   ffmpeg -r 20 -i temp/bloch_%01d.png -crf 10 -b:v 1800 bloch.mp4
```

Comment:
This works for me, but I don't know what `-crf` does or how it fixes the issues or what the value `10` is. It would be good to understand this. Do you perhaps know? I'm a bit worried that `-crf 10` only fixes the issue as a side effect and that there is a more direct / correct fix (I could be wrong, of course).

Reply:
From my preliminary understanding, the `crf` option could be used to control the quality of the video, and it looks like it is recommended in most scenarios, as I found [here](https://trac.ffmpeg.org/wiki/Encode/H.264). From the output of the `ffmpeg` command on my local machine, it seems that I have the `libx264` encoder, and I found [here](https://slhck.info/video/2017/03/01/rate-control.html) that there could be several options (depending on the type of encoder used) for controlling the quality, of which `crf` seems to be used widely.

Removing the `crf` parameter gives a blank video, as you had pointed out, and hence there might have been no default value set already. As a result, the `crf` parameter could prove to be a safe choice to render the video successfully on all machines and be a plausible solution?

Could you let me know your thoughts on this? I would try to figure out if the blank video could be solved by tinkering with something else, for eg: controlling the image size instead.

Reply:
I don't know the answer unfortunately. I would just like to know what is happening before merging something we don't understand.

Reply:
Specifying a hard-coded CRF without fixing the encoder sounds a bit dangerous to me.  When I've historically used `ffmpeg` to make videos from still images, I'd tend to specify something like
```bash
ffmpeg -i 'files_%03d.png' -c:v libx264 -vf 'fps=25,format=yuv420p' bloch.mp4
```
and let just have the encoding sort out the rest.  (I don't know what was causing the blank image in this case, so this command may not work first try here.) For these purposes I wouldn't worry much about fixing any bitrate or quality in the encoding - it would be much more efficient for the user to control those features by setting the DPI/size of the _images_, rather than producing them at high resolution and downsampling in the video conversion.

Reply:
I tried to investigate by tweaking a couple of parameters and came up with the following:

- Considering that `ffmpeg` is intelligent in figuring out the relevant settings (which are generally given by the switch arguments), I thought it would be a good choice to keep the number of arguments to `ffmpeg` at the minimum. This eases readability without having to think much about all the arguments and at the same time allowing `ffmpeg` to figure out the appropriate arguments based on the intrinsic parameters of the device the user uses.
- A quick analysis showed that the major problem in the command is in the bitrate parameter (`-b`) which is just 1800, i.e. 1800 bits per second, an excessively low value for videos.

For a comparative analysis, below is a screenshot by using this command (10k bps):

```
ffmpeg -i temp/bloch_%01d.png -b:v 10000 bloch.mp4
```
![10k_bitrate](https://user-images.githubusercontent.com/68844397/118362830-f1c3a580-b5ae-11eb-9546-821f1b266cbb.png)

As can be seen, even using 10000 gives a blurred video.

- Combining the above two points, a plausible solution might be to get rid of all these arguments and instead use:

```
ffmpeg -i temp/bloch_%01d.png bloch.mp4
```

which renders the video by a "default" bitrate and other relevant parameters which leads to an acceptable video quality.

![default_bitrate](https://user-images.githubusercontent.com/68844397/118362953-a8278a80-b5af-11eb-9662-e1a8b0eb6010.png)


**Advantages**
There would be no need to do something like `-c:v libx264` since by getting rid of such arguments, we let `ffmpeg` decide the encoder. This could immediately imply that this command becomes more customizable.

It would be really good to get your thoughts on this @jakelishman @hodgestar !

Reply:
10kbps is also incredibly low for videos in general, so it's not surprising that it was blurry as anything.  I haven't tried it, but I'd estimate a sensible .png size for a single frame is ~5kB (kilo*bytes*), which is 40kb (kilo*bits*).  You only have enough bitrate to fit a single .png in _four_ seconds of video at 10kbps, and video compression of 20fps with a keyframe every ~2s simply can't handle that without killing the quality completely.

The ffmpeg defaults are the most sensible way to do it, for sure.  We're not an ffmpeg tutorial, after all.